src/internal_tricks.rs

```diff
@@ -35,15 +35,45 @@ macro_rules! pyo3_exception {
 #[derive(Debug)]
 pub(crate) struct NulByteInString(pub(crate) &'static str);
 
+#[derive(Copy, Clone)]
+pub(crate) enum MaybeLeaked {
+    Static(*const c_char),
+    Leaked(*mut c_char),
+}
+
+impl MaybeLeaked {
+    pub(crate) fn as_ptr(&self) -> *const c_char {
+        match *self {
+            Self::Static(ptr) => ptr,
+            Self::Leaked(ptr) => ptr as *const c_char,
+        }
+    }
+}
+
 pub(crate) fn extract_cstr_or_leak_cstring(
     src: &'static str,
     err_msg: &'static str,
-) -> Result<&'static CStr, NulByteInString> {
-    CStr::from_bytes_with_nul(src.as_bytes())
-        .or_else(|_| {
-            CString::new(src.as_bytes()).map(|c_string| &*Box::leak(c_string.into_boxed_c_str()))
-        })
-        .map_err(|_| NulByteInString(err_msg))
+) -> Result<MaybeLeaked, NulByteInString> {
+    let bytes = src.as_bytes();
+    let nul_pos = memchr::memchr(0, bytes);
```

Comment:
Otherwise we need to traverse `bytes` twice; both of those functions check for internal nul bytes.

Reply:
I understand that. I would say that the price of a new dependency (`memchr` was only a dev-dependency until now) and more unsafe blocks is relatively steep for that. I also suspect that the compiler might just be able to optimize out the redundant checks in the safe alternatives.

(As an aside, the change is somewhat orthogonal to the topic of the PR. It might be good to propose it separately to avoid derailing the fix for the memory leak. But then, this is probably not too convincing coming from the one doing the derailing in the first place.)

Reply:
How about this alternative (pseudocode):

```rust
let bytes = src.as_bytes();
if !bytes.is_empty() && bytes[-1] == '\0' {
    CStr::from_bytes_with_nul(bytes)
} else {
    CString::new(bytes)
}
```

Checking for the trailing nul doesn't need an external dependency. We can then branch on the leak-or-not operation and we know that any errors must be due to internal nuls.